### ISSUE TITLE
ggplot methods for plotting 2-dimensional HPD regions

### ISSUE THREAD

**Issue by crsh** [mjskay/tidybayes#214]
Hi there,

great package. I was wondering whether you might be interested in adding a new `ggplot2` stat and geom.

I recently wanted to plot a two-dimensional HPD region for the posterior predictions of my model. Because of my bounded outcome variable, I was unhappy with the normal approximation created with `stat_ellipse()`. Merging `emdbook::HPDregionplot()` and `StatContour()`, I came up with the following solution based on `MASS::kde2d()`:

~~~r
#' Contours of a HDP region estimate
#'
#' @param prob Numeric. Probability level of the HDP region.
#' @inheritParams ggplot2::stat_density_2d
#' @inheritParams MASS::kde2d
#' @param ...
#'
#' @return
#' @export

stat_hpd_2d <- function(mapping = NULL, data = NULL, geom = "polygon",
                        position = "identity", na.rm = FALSE, show.legend = NA,
                        inherit.aes = TRUE, n = 100, prob = 0.95, ...) {
  ggplot2::layer(
    stat = StatHPDContour, data = data, mapping = mapping, geom = geom,
    position = position, show.legend = show.legend, inherit.aes = inherit.aes,
    params = list(na.rm = na.rm, n = n, prob = prob, ...)
  )
}

StatHPDContour <- ggplot2::ggproto(
  "hpd_2d"
  , Stat
  , compute_group = function (data, scales, na.rm = FALSE, h = NULL,
                              n = 100, prob = 0.95)
  {
    if (is.null(h)) {
      h <- c(MASS::bandwidth.nrd(data$x), MASS::bandwidth.nrd(data$y))
    }
    dens <- MASS::kde2d(data$x, data$y, h = h, n = n,
                        lims = c(scales$x$dimension(), scales$y$dimension()))
    df <- data.frame(expand.grid(x = dens$x, y = dens$y), z = as.vector(dens$z))
    df$group <- data$group[1]

    dx <- diff(dens$x[1:2])
    dy <- diff(dens$y[1:2])
    sz <- sort(dens$z)
    c1 <- cumsum(sz) * dx * dy

    breaks <- sapply(prob, function(x) {
      withCallingHandlers(
        stats::approx(c1, sz, xout = 1 - x)$y
        , warning = function(w) {
          if (grepl("collapsing to unique 'x' values", w$message))
            invokeRestart("muffleWarning")
        }
      )
    })

    ggplot2::StatContour$compute_panel(df, scales, breaks = breaks)
  }
  , required_aes = c("x", "y")
)
~~~

Here's an example:

~~~r
library("ggplot2")
ggplot(data = iris, aes(x = Sepal.Length, y = Sepal.Width, color = Species)) +
    stat_hpd_2d(aes(fill = Species), prob = 0.8, alpha = 0.1, linetype = "22", size = 0.3) +
    geom_point(size = 3, alpha = 0.6)
~~~

![stat_hpd](https://user-images.githubusercontent.com/2195485/70143627-7a40d580-169c-11ea-81ee-b3814291a690.png)

If this is something you'd be interested in, I'd be happy to expand this into a PR (add a geom, create aliases such as `stat_hpd2d()`, etc.). What do you think?

**Comment by mjskay** [comment]
This could be useful! I'm not totally clear on the difference between this and `stat_density_2d`, is it the parameterization in terms of desired mass inside the contour? (it's not immediately clear to me how to do that with `stat_density_2d` / `geom_contour` but I haven't played with them much).

**Comment by crsh** [comment]
Good question. I probably should have laid that out more clearly.

#### Why not `stat_density_2d()`?

To control the resulting contours `stat_density_2d()` accepts only the `bins` argument, which is [passed to](https://github.com/tidyverse/ggplot2/blob/913e936ce4abfd933897ca9e5425dea76bd7d596/R/stat-density-2d.r#L80) `StatContour`. This controls how many bins are drawn, but as far as I understand, those do not readily correspond to enclosed probability masses. While `StatContour` [accepts](https://github.com/tidyverse/ggplot2/blob/913e936ce4abfd933897ca9e5425dea76bd7d596/R/stat-contour.r#L102) a `breaks` argument, which would grant greater control, this is not accessible from `stat_density_2d()`.

#### Why not `stat_contour()`?

`stat_contour()` requires a `z` aesthetic. So in order to do the above, it would be necessary to first call, for example, `MASS::kde2d()` manually (for each group), then call `stat_contour()`, and pass the `breaks` argument, which seems inconvenient (from [looking at the code](https://github.com/tidyverse/ggplot2/blob/913e936ce4abfd933897ca9e5425dea76bd7d596/R/stat-contour.r#L105) it seems `breaks` is not utilized; this is probably a bug, which I will try to address at the `ggplot2` repository). Moreover,`StatContour` does not accept breaks in units of enclosed probability mass, which I find very useful to visualize estimate uncertainty.

**Comment by mjskay** [comment]
Makes sense. Okay yeah, I'd be happy to have something like this. I think it would be good if it was similar to stat_lineribbon in that:

- it can provide multiple levels a vector parameter of desired probability masses

- all regions with the same mass are drawn below all other regions with less mass (e.g. if the user asks for 95, 80, and 50% regions, all 95% regions are drawn, then 80% regions, then 50% regions). `geom_lineribbon` currently does this (in a fairly hackish way that could be better I think) so that if you use transparency and you have multiple different groups with some overlap it is easier to see both groups. This gives an interleaving effect that you can see in this example from `vignette("tidy-brms")`:

![image](https://user-images.githubusercontent.com/6345019/70569384-d9ed2400-1b67-11ea-8af7-2f43956aec44.png)

- (possibly) might be nice to be able to do a point summary as well. Having a point summary in there would potentially allow for easy uncertainty-including scatterplots of parameters, which could be cool.

Happy to accept a PR on this if you want to take it on. Name-wise I'm not too keen on "HPD" (there's no reason you have to use it on a posterior) so maybe something that echos "HDI" or even a non-acronym form (like "geom_density_region") might be good.

**Comment by crsh** [comment]
Sounds good to me. I'm not sure how soon I'll get around to it, but I'll give it a go.

**Comment by mjskay** [comment]
Great, thanks!

**Comment by mjskay** [comment]
Moving this issue to {ggdist} as all the geoms and stuff have been moved there.
